streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java

```diff
@@ -240,27 +185,33 @@ private void emitNonJoinedOuterRecords(
                     // There might be an outer record for the other joinSide which window has not closed yet
                     // We rely on the <timestamp><left/right-boolean><key> ordering of KeyValueIterator
                     final long outerJoinLookBackTimeMs = getOuterJoinLookBackTimeMs(timestampedKeyAndJoinSide);
-                    if (sharedTimeTracker.minTime + outerJoinLookBackTimeMs + joinGraceMs >= sharedTimeTracker.streamTime) {
+                    if (sharedTimeTracker.minTime + outerJoinLookBackTimeMs + joinGraceMs
+                            >= sharedTimeTracker.streamTime) {
                         if (timestampedKeyAndJoinSide.isLeftSide()) {
-                            outerJoinLeftWindowOpen = true; // there are no more candidates to emit on left-outerJoin-side
+                            outerJoinLeftWindowOpen =
+                                    true; // there are no more candidates to emit on left-outerJoin-side
                         } else {
-                            outerJoinRightWindowOpen = true; // there are no more candidates to emit on right-outerJoin-side
+                            outerJoinRightWindowOpen =
+                                    true; // there are no more candidates to emit on right-outerJoin-side
                         }
                         // We continue with the next outer record
                         continue;
                     }
-                    
+
                     final K key = timestampedKeyAndJoinSide.getKey();
-                    final LeftOrRightValue<V1, V2> leftOrRightValue = next.value;
-                    final VOut nullJoinedValue = getNullJoinedValue(key, leftOrRightValue);
+                    final LeftOrRightValue<VL, VR> leftOrRightValue = next.value;
+                    final VThis thisValue = getThisValue(leftOrRightValue);
+                    final VOther otherValue = getOtherValue(leftOrRightValue);
+                    final VOut nullJoinedValue = joiner.apply(key, thisValue, otherValue);
                     context().forward(
-                        record.withKey(key).withValue(nullJoinedValue).withTimestamp(timestamp)
+                            record.withKey(key).withValue(nullJoinedValue).withTimestamp(timestamp)
                     );
```

Comment:
nit: This could be brought out to a method which depends only on `record` and `next`.